File: android/src/main/kotlin/com/bitmovin/player/PlayerViewMethod.kt
package com.bitmovin.player

import android.content.Context
import android.view.View
import com.bitmovin.core.Channels
import com.bitmovin.core.data.Methods
import com.bitmovin.player.api.Player
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

class PlayerViewMethod(
    context: Context,
    messenger: BinaryMessenger,
    id: Int,
) : MethodChannel.MethodCallHandler, PlatformView {
    @Suppress("unused")
    private val tag: String = this::class.java.simpleName
    private var view: PlayerView
    private var playerId: String? = null

    init {
        MethodChannel(
            messenger,
            "${Channels.PLAYER_VIEW}-$id",
        ).apply {
            this.setMethodCallHandler(this@PlayerViewMethod)
        }
        view = PlayerView(context)
    }

    private fun setPlayer(player: Player) {
        view.player = player
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val params = call.arguments as Map<*, *>
        playerId = params["playerId"] as String
        when (call.method) {
            Methods.BIND_PLAYER -> {
                PlayerManager.players[playerId]?.let {
                    setPlayer(it)
                }
            }
            Methods.UNBIND_PLAYER -> {
                dispose()
            }
        }
    }

    override fun getView(): View {
        return view
    }

    override fun dispose() {
        view.player = null
        playerId?.let {
            PlayerManager.destroy(it)
        }
    }
}

File: ios/Classes/PlayerViewMethod.swift
import Foundation
import Flutter
import BitmovinPlayer

// TODO: Refactor this class
class PlayerViewMethod: NSObject, FlutterPlatformView {
    private var viewIdentifier: Int64?
    private var frame: CGRect?
    private var messenger: FlutterBinaryMessenger?
    private var arguments: Any?

    private var _view: UIView = UIView()
    private var _playerView: PlayerView?
    private var _player: Player?
    private var _methodChannel: FlutterMethodChannel?

    init(
        viewIdentifier: Int64,
        frame: CGRect,
        arguments: Any?,
        binaryMessenger messenger: FlutterBinaryMessenger
    ) {
        self._view = UIView()
        super.init()

        self.viewIdentifier = viewIdentifier
        self.frame = frame
        self.messenger = messenger
        self.arguments = arguments
        self._methodChannel = FlutterMethodChannel(
            name: Channels.playerView + "-\(String(describing: viewIdentifier))",
            binaryMessenger: messenger
        )
        self._methodChannel?.setMethodCallHandler(self.handleMethodCall)
    }

    private func createPlayerView(playerId: String, playerViewConfig: PlayerViewConfig?) {
        _playerView = PlayerView(
            player: PlayerManager.shared.players[playerId]!,
            frame: UIView().bounds,
            playerViewConfig: playerViewConfig ?? PlayerViewConfig()
        )
        _playerView?.autoresizingMask = [.flexibleHeight, .flexibleWidth]
        _view.addSubview(_playerView!)
    }

    func view() -> UIView {
        _view.backgroundColor = UIColor.black
        return _view
    }

    private func handleMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any?] else { return }

        if call.method == Methods.bindPlayer, let playerId = args["playerId"] as? String {
            createPlayerView(playerId: playerId, playerViewConfig: nil)
        }
    }
}
